### PR TITLE
Fix dupe handling in Splay ACLs: src, dst, http_status, etc.

### PR DESCRIPTION
**Pull request by rousskov**
Squid was dangerously mishandling duplicate[^1] values in ACLs with
Splay tree storage: src, dst, localip, http_status, dstdomain,
srcdomain, and ssl::server_name. These problems were all rooted in the
same old code but had diverged across two ACL groups, as detailed in the
corresponding sections below. Fortunately, all known problems were
accompanied with Squid cache.log WARNINGs: Squid configurations that do
not emit cited or similar warnings are not affected by these bugs.

[^1]: Squid Splay trees can only store unique values. Uniqueness is
defined in terms of a configurable comparison function that returns zero
when the two values are "not unique" (i.e. are considered to be
"duplicated" or "equal" in that Splay context). Those two "equal" values
may differ a lot in other contexts! For example, the following two
status code ranges are equal from acl_httpstatus_data::compare() point
of view, but are obviously very different from an admin and http_access
rules points of view: 200-200 and 200-299.

### Group 1: src, dst, localip, and http_status ACLs

These ACLs ignored (i.e. never matched) configured ACL values. They also
gave wrong "remove X" advice and incorrectly classified values as being
subranges or subsets:

    Processing: acl testA http_status 200 200-300
    WARNING: '200-300' is a subrange of '200'
    WARNING: because of this '200-300' is ignored ...
    
    Processing: acl testB http_status 300-400 200-300
    WARNING: '200-300' is a subrange of '300-400'
    WARNING: because of this '200-300' is ignored ...
    WARNING: You should probably remove '300-400' from the ACL...

    Processing: acl testC src 10.0.0.1  10.0.0.0-10.0.0.255
    WARNING: (B) 10.0.0.1 is a subnetwork of (A) 10.0.0.0-10.0.0.255
    WARNING: because of this 10.0.0.0-10.0.0.255 is ignored...

    Processing: acl testD src 10.0.0.0-10.0.0.1  10.0.0.1-10.0.0.255
    WARNING: (A) 10.0.0.1-10.0.0.255 is a subnetwork of (B) 10.0.0.0-...
    WARNING: because of this 10.0.0.1-10.0.0.255 is ignored...
    WARNING: You should probably remove 10.0.0.1-10.0.0.255 from the ACL

Since 2002 commit 96da6e8, IP based ACLs like src, dst, and localip
eliminate duplicates[^1] among configured ACL values. The elimination
code was buggy since inception. Those bugs were later duplicated in
http_status code (2005 commit a0ec9f6). This change fixes those bugs.

To correctly eliminate duplicates when facing two (fully or partially)
overlapping ranges -- new A and old B -- we must pick the right
corrective action depending on the kind of overlap:

    * A is a subrange of B: Ignore fully duplicated new range A. Keep B.
    * B is a subrange of A: Remove fully duplicated old range B. Add A.
    * A partially overlaps with B: Add a union of A and B. Remove B.
    
Both acl_httpstatus_data::compare() and acl_ip_data::NetworkCompare()
mishandled the last two cases because these functions effectively
implemented the following buggy logic instead:

    - in all three cases: Ignore new range A. Keep B.

Their WARNINGs also suggested wrong corrective actions in two mishandled
cases (see the last WARNING lines in testB and testD output above).

### Group 2: dstdomain, srcdomain, and ssl::server_name ACLs

    Processing: acl testE dstdomain .example.com example.com
    ERROR: '.example.com' is a subdomain of 'example.com'
    ERROR: You need to remove '.example.com' from the ACL named 'testE'

The above 2002 commit 96da6e8 bugs stopped affecting domain-based ACLs
(i.e. dstdomain, srcdomain, and ssl::server_name) in 2011 commit 14e563c
that adjusted aclDomainCompare() to self_destruct() in problematic
cases. However, that adjustment emitted wrong advice and incorrect
subdomain classification in cases where strlen() checks do not work for
determining which of the two configured ACL values should be removed
(see testE ERRORs above).
    
This change improves that handling by replacing the call to
self_destruct() with proper duplicate resolution code (and fixing
cache.log messages). Internal Splay limitations and resource waste
aside, duplicate values do not invalidate an ACL -- an ACL with
duplicates could match as expected. It may be difficult for some admins
to avoid duplication, especially when ACL values come from multiple
sources. Squid should continue to warn about duplicates (because they
waste resources and may indicate a deeper problem), but killing Squid or
otherwise rejecting ACLs with duplicates is bad UX.

N.B. Domain-based ACLs use sets of values rather than "ranges" discussed
in "Group 1" section, but domain sets follow the same basic principles.

### All of the above seven ACLs

The problems in both ACL groups were fixed by factoring out "insert ACL
values while correctly handling duplicates" algorithm (see the three
bullets in section "Group 1") into Acl::SplayInserter::Merge() function.

This change also fixes memory leaks associated with ignored ACL values.

Also added test-suite/test-squid-conf.sh support for matching multiple
error messages. Without this, testing a variety of closely-related cases
requires creating lots of test configuration files (two per test case),
increasing noise and, more importantly, making it difficult to handle
related test cases as one coherent collection. The new "expect-messages"
feature is barely sufficient for testing these changes, but we are now
at (or perhaps even beyond) the limit of what can be reasonably done
using shell scripts and test instruction files: The next step is to
convert instruction files (and likely some test cases themselves!) to
scripts written in Perl or a better language.

### CODE REVIEW

File: src/acl/HttpStatus.cc [hunk 1]
```diff
@@ -132,13 +147,13 @@ aclMatchHTTPStatus(Splay<acl_httpstatus_data*> **dataptr, const Http::StatusCode
 static int
 aclHTTPStatusCompare(acl_httpstatus_data * const &a, acl_httpstatus_data * const &b)
 {
-    if (a->status1 < b->status1)
-        return 1;
+    if (a->status2 < b->status1)
+        return 1; // the entire range a is to the left of range b
```
Comment (rousskov) [inline src/acl/HttpStatus.cc:151]: IMO, it would be more logical to return `-1` here (and `+1` below), but I decided not to change this (correct) old code. I did replace `a->status1` with `a->status2` above to make this code work for `a` values representing multiple status codes. For single-status ranges that this function used to handle (and still does, in addition to handling multi-status ranges), there is no difference because status1 is set to status2 for single-status ranges.

File: src/acl/SplayInserter.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    /// function used later by ACL::match().
+    /// \retval 0 The two values overlap. Here, two values overlap if they are
+    /// identical, if one contains all values from another, and if one contains
+    /// at least one value from another.
+    static int Compare(const Value &, const Value &);
+
+    /// whether the set of values matched by `a` contains the entire set of
+    /// values matched by `b`, including cases where `a` is identical to `b`
+    /// \prec The two values overlap: Compare(a, b) == 0
+    static bool AcontainsEntireB(const Value &a, const Value &b);
```
Comment (rousskov) [inline src/acl/SplayInserter.h:51]: I do not like the name of this function, and it violates Squid naming style, but I cannot think of a better variant.

File: test-suite/squidconf/bad-acl-src-dupe.conf [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+acl test14 src 127.0.0.0-127.0.0.128  127.0.0.128-127.0.0.255
+
+acl test15 src 10.0.0.1-10.0.0.128 10.0.0.0-10.0.0.1 10.0.0.128-10.0.0.255
+
+acl test25 dst 127.0.0.0-127.0.0.128/32  127.0.0.128-127.1.0.255
+
+acl test36 dst 127.0.0.1-127.0.0.128  127.0.0.0-127.1.0.0/16
+acl test37 dst 127.0.0.0-127.1.0.0/16 127.0.0.1-127.0.0.128
+
+# TODO: make configurable depending on USE_IPV6
```
Comment (rousskov) [inline test-suite/squidconf/bad-acl-src-dupe.conf:21]: I know we have code to address this TODO. We will add that code in this PR if others insist. Otherwise, we will do a small dedicated PR after asking whether others would prefer to

* split such test files into IPv4 and IPv6 ones OR
* keep all tests together but skip (all) tests in no-IPv6 cases.

File: src/acl/SplayInserter.h (hunk 2 above)
Comment (eduard-bagdasaryan) [reply src/acl/SplayInserter.h:51]: Can we simply name it 'IsSuperset()', meaning that a is a superset of b? Also we could have a paired IsSubset(a, b) and use it in SplayInserter<DataValue>::Merge() instead of (a little confusing) AcontainsEntireB() with reversed arguments.

File: src/acl/SplayInserter.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+} // namespace Acl
+
+template <class DataValue>
+void
+Acl::SplayInserter<DataValue>::Merge(Splay<Value> &storage, Value &&newItem)
+{
+    const auto comparator = &SplayInserter<Value>::Compare;
+    while (const auto oldItemPointer = storage.insert(newItem, comparator)) {
+        const auto oldItem = *oldItemPointer;
```
Comment (eduard-bagdasaryan) [inline src/acl/SplayInserter.h:70]: The assertion below implies that oldItem is a pointer. Than we probably need something like `static_assert(std::is_pointer<Value>::value);`

File: src/acl/Ip.cc [hunk 5]
```diff
@@ -438,8 +472,7 @@ ACLIP::parse()
             /* pop each result off the list and add it to the data tree individually */
             acl_ip_data *next_node = q->next;
             q->next = nullptr;
-            if (!data->find(q,acl_ip_data::NetworkCompare))
-                data->insert(q, acl_ip_data::NetworkCompare);
+            Acl::SplayInserter<acl_ip_data*>::Merge(*data, std::move(q));
```
Comment (eduard-bagdasaryan) [inline src/acl/Ip.cc:475]: Since q is a pointer - can we just pass it by value instead of this complexity with std::move()?

File: test-suite/squidconf/bad-acl-http-status-dupe.conf [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+##
+
+acl test11 http_status 400 400
+acl test12 http_status 407 400 407
+acl test13 http_status 200 200-300
+
+acl test14 http_status 300-399
+acl test14 http_status 304
+
+acl test21 http_status 100-300 350 200-400
```
Comment (eduard-bagdasaryan) [inline test-suite/squidconf/bad-acl-http-status-dupe.conf:15]: What about 'reversed' ranges like 300-100? It looks like Squid treats this range as a single '300' value (regardless of PR changes), which is not intuitive, I think. It does not warn about them either. Similarly for ip ranges.

File: src/acl/SplayInserter.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+private:
+    /// SplayInserter users are expected to specialize all or most of the static
+    /// methods below. Most of these methods have no generic implementation.
+
+    /// A Splay::SPLAYCMP function for comparing parsed ACL parameter values.
+    /// This function must work correctly with all valid ACL parameter values,
+    /// including those representing sets or ranges. The order specified by this
+    /// function must be the same as the order specified by the SPLAYCMP
+    /// function used later by ACL::match().
+    /// \retval 0 The two values overlap. Here, two values overlap if they are
```
Comment (eduard-bagdasaryan) [inline src/acl/SplayInserter.h:43]: What about other return values?

File: src/acl/SplayInserter.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            DestroyValue(newItem);
+            return;
+        }
+
+        if (AcontainsEntireB(newItem, oldItem)) {
+            debugs(28, DBG_PARSE_NOTE(DBG_IMPORTANT), "WARNING: Ignoring earlier " << oldItem << " because it is covered by " << newItem <<
+                   Debug::Extra << "advice: Remove value " << oldItem << " from the ACL named " << AclMatchedName);
+            storage.remove(oldItem, comparator);
+            DestroyValue(oldItem);
+            continue; // still need to insert newItem (and it may conflict with other old items)
```
Comment (eduard-bagdasaryan) [inline src/acl/SplayInserter.h:85]: We could remove both 'continue' and add an 'else' below (or keep as-is).

**Review by eduard-bagdasaryan — COMMENTED** [review]
Partially reviewed.
Also tested with the available tests.

File: test-suite/squidconf/bad-acl-dstdomain-dupe.conf [hunk 9]
```diff
@@ -0,0 +1,19 @@
+## Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+acl test11 dstdomain .example.com example.com
```
Comment (eduard-bagdasaryan) [inline test-suite/squidconf/bad-acl-dstdomain-dupe.conf:8]: Also don't we need some tests checking for false positives, e.g:
```
acl test11 dstdomain a.example.com b.example.com
```

File: test-suite/test-squid-conf.sh [hunk 10]
```diff
@@ -43,6 +47,121 @@ expectMessage()
...
+
+# Checks that each of the extended regexes (in the given file) matches line(s)
+# in the given Squid log file. A log line can match at most once.
+matchEachRegex()
+{
+    local regexFilename="$1"
+    local errLog="$2"
+
+    local errorLogRemaining="$errorLog.unmatched";
+    local errorLogNext="$errorLog.next";
```
Comment (eduard-bagdasaryan) [inline test-suite/test-squid-conf.sh:127]: It looks like this file is not listed among CLEANFILES.

File: src/acl/SplayInserter.h (hunk 7 above)
Comment (rousskov) [reply src/acl/SplayInserter.h]: Refactored in commit 030069f.

File: src/acl/SplayInserter.h (hunk 2 above)
Comment (rousskov) [reply src/acl/SplayInserter.h]: Switched to IsSubset() in d8bb057 under the assumption that my fear of folks mismatching which subset() argument is a subset is unfounded.

File: src/acl/SplayInserter.h (hunk 4 above)
Comment (rousskov) [reply src/acl/SplayInserter.h:72]: I disagree that assert(x) implies that x is (or should be) a pointer, but I removed that (arguably stale) assertion in commit e155f42 because the current Merge() code no longer dereferences oldItem (explicitly).

**Comment by rousskov** [comment]
This became official PR 1632.
